### DISCUSSION TITLE
Error  while setting up virtual IOT environment

### DISCUSSION THREAD

**Discussion by InamulRehman** [microsoft/IoT-For-Beginners#359] [Q&A]
hello! 
i was setting up virtual environment for IOT but getting error while running counterfit command.
I have followed all instructions in Virtual single-board computer very precisely and while running the command **pip install CounterFit**  i got this kind of error :
![Screenshot 2022-02-05 172745](https://user-images.githubusercontent.com/80153613/152642170-2dd82ca0-2a43-4264-b2ed-1067c30f2a19.png)
![Screenshot 2022-02-05 172831](https://user-images.githubusercontent.com/80153613/152642194-6cbe0ff4-23ff-4505-aaf0-eccb88d256bb.png)
But i proceded with the next folllowing commands **pip install counterfit-connection** , 
**pip install counterfit-shims-grove**  which get run successfully. 
After that running the **counterfit** i got
![Screenshot 2022-02-05 173320](https://user-images.githubusercontent.com/80153613/152642312-50fcaf9a-8234-45e4-9b37-b9ba68bd1c0f.png)

I am not getting what the problem is ?
So please suggest me the solution to avoid the problem and be able to run counterfit app on my system.
Thanks!

**Comment by jimbobbennett** [comment]
This was an issue where one of the packages used doesn't work with Python 3.10. It's now been fixed in CounterFit, so can you try again please? Should now work.

**Reply by InamulRehman** [reply]
![err](https://user-images.githubusercontent.com/80153613/175113976-a148f0a4-9376-4351-b6dc-1036e05215b5.JPG)


Counterfit app not getting installed. Above is the error log.

**Reply by jimbobbennett** [reply]
Can you copy/past the entire output rather than just a screenshot. I can't see an actually error message on that screen shot.

I'm guessing this is Windows, but are you still using PowerShell, or is this in the command prompt?

**Comment by InamulRehman** [comment]
Al right i will try again . And i will show you results..

**Comment by InamulRehman** [comment]
I was using windows *power shell.*

**Reply by jimbobbennett** [reply]
Can you paste the entire error message please

**Comment by InamulRehman** [comment]
(.venv) PS D:\nightlight> pip install CounterFit
Collecting CounterFit
  Using cached CounterFit-0.1.4.dev6-py3-none-any.whl (119 kB)
Collecting beautifulsoup4==4.9.3
  Using cached beautifulsoup4-4.9.3-py3-none-any.whl (115 kB)
Collecting Flask-SocketIO==5.2.0
  Using cached Flask_SocketIO-5.2.0-py3-none-any.whl (17 kB)
Collecting lxml==4.6.4
  Using cached lxml-4.6.4.tar.gz (3.2 MB)
Requirement already satisfied: Flask==2.1.2 in
d:\nightlight\.venv\lib\python3.9\site-packages (from CounterFit) (2.1.2)
Collecting eventlet==0.33.1
  Using cached eventlet-0.33.1-py2.py3-none-any.whl (226 kB)
Requirement already satisfied: soupsieve>1.2 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
beautifulsoup4==4.9.3->CounterFit) (2.3.2.post1)
Requirement already satisfied: dnspython>=1.15.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
eventlet==0.33.1->CounterFit) (2.2.1)
Requirement already satisfied: six>=1.10.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
eventlet==0.33.1->CounterFit) (1.16.0)
Requirement already satisfied: greenlet>=0.3 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
eventlet==0.33.1->CounterFit) (1.1.2)
Requirement already satisfied: Werkzeug>=2.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask==2.1.2->CounterFit) (2.1.2)
Requirement already satisfied: importlib-metadata>=3.6.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask==2.1.2->CounterFit) (4.11.4)
Requirement already satisfied: itsdangerous>=2.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask==2.1.2->CounterFit) (2.1.2)
Requirement already satisfied: Jinja2>=3.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask==2.1.2->CounterFit) (3.1.2)
Requirement already satisfied: click>=8.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask==2.1.2->CounterFit) (8.1.3)
Requirement already satisfied: python-socketio>=5.0.2 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Flask-SocketIO==5.2.0->CounterFit) (5.6.0)
Requirement already satisfied: colorama in
d:\nightlight\.venv\lib\python3.9\site-packages (from
click>=8.0->Flask==2.1.2->CounterFit) (0.4.5)
Requirement already satisfied: zipp>=0.5 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
importlib-metadata>=3.6.0->Flask==2.1.2->CounterFit) (3.8.0)
Requirement already satisfied: MarkupSafe>=2.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
Jinja2>=3.0->Flask==2.1.2->CounterFit) (2.1.1)
Requirement already satisfied: bidict>=0.21.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
python-socketio>=5.0.2->Flask-SocketIO==5.2.0->CounterFit) (0.22.0)
Requirement already satisfied: python-engineio>=4.3.0 in
d:\nightlight\.venv\lib\python3.9\site-packages (from
python-socketio>=5.0.2->Flask-SocketIO==5.2.0->CounterFit) (4.3.2)
Using legacy 'setup.py install' for lxml, since package 'wheel' is not
installed.
Installing collected packages: lxml, Flask-SocketIO, eventlet,
beautifulsoup4, CounterFit
    Running setup.py install for lxml ... error
    ERROR: Command errored out with exit status 1:
     command: 'D:\nightlight\.venv\bin\python3.exe' -u -c 'import io, os,
sys, setuptools, tokenize; sys.argv[0] =
'"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-install-nbpaxmto\\lxml_f6d5c4e067ed43e8bf4a1f3a1db7bbca\\setup.py'"'"';
__file__='"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-install-nbpaxmto\\lxml_f6d5c4e067ed43e8bf4a1f3a1db7bbca\\setup.py'"'"';f
= getattr(tokenize, '"'"'open'"'"', open)(__file__) if
os.path.exists(__file__) else io.StringIO('"'"'from setuptools import
setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"',
'"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))'
install --record
'C:\Users\Dell\AppData\Local\Temp\pip-record-7xh23b80\install-record.txt'
--single-version-externally-managed --compile --install-headers
'D:\nightlight\.venv\include\site\python3.9\lxml'
         cwd:
C:\Users\Dell\AppData\Local\Temp\pip-install-nbpaxmto\lxml_f6d5c4e067ed43e8bf4a1f3a1db7bbca\
    Complete output (97 lines):
    Building lxml version 4.6.4.
    Building without Cython.
    Building against pre-built libxml2 andl libxslt libraries
    running install
    running build
    running build_py
    creating build
    creating build\lib.mingw_x86_64-3.9
    creating build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\builder.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\cssselect.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\doctestcompare.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\ElementInclude.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\pyclasslookup.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\sax.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\usedoctest.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\_elementpath.py -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\__init__.py -> build\lib.mingw_x86_64-3.9\lxml
    creating build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\__init__.py ->
build\lib.mingw_x86_64-3.9\lxml\includes
    creating build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\builder.py -> build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\clean.py -> build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\defs.py -> build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\diff.py -> build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\ElementSoup.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\formfill.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\html5parser.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\soupparser.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\usedoctest.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\_diffcommand.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\_html5builder.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\_setmixin.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    copying src\lxml\html\__init__.py ->
build\lib.mingw_x86_64-3.9\lxml\html
    creating build\lib.mingw_x86_64-3.9\lxml\isoschematron
    copying src\lxml\isoschematron\__init__.py ->
build\lib.mingw_x86_64-3.9\lxml\isoschematron
    copying src\lxml\etree.h -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\etree_api.h -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\lxml.etree.h -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\lxml.etree_api.h -> build\lib.mingw_x86_64-3.9\lxml
    copying src\lxml\includes\c14n.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\config.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\dtdvalid.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\etreepublic.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\htmlparser.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\relaxng.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\schematron.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\tree.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\uri.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xinclude.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xmlerror.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xmlparser.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xmlschema.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xpath.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\xslt.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\__init__.pxd ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\etree_defs.h ->
build\lib.mingw_x86_64-3.9\lxml\includes
    copying src\lxml\includes\lxml-version.h ->
build\lib.mingw_x86_64-3.9\lxml\includes
    creating build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources
    creating build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\rng
    copying src\lxml\isoschematron\resources\rng\iso-schematron.rng ->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\rng
    creating build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl
    copying src\lxml\isoschematron\resources\xsl\RNG2Schtrn.xsl ->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl
    copying src\lxml\isoschematron\resources\xsl\XSD2Schtrn.xsl ->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl
    creating
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_abstract_expand.xsl
->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_dsdl_include.xsl
->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_schematron_message.xsl
->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_schematron_skeleton_for_xslt1.xsl
->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_svrl_for_xslt1.xsl
->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    copying
src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\readme.txt ->
build\lib.mingw_x86_64-3.9\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
    running build_ext
    building 'lxml.etree' extension
    creating build\temp.mingw_x86_64-3.9
    creating build\temp.mingw_x86_64-3.9\src
    creating build\temp.mingw_x86_64-3.9\src\lxml
    gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall
-march=x86-64 -mtune=generic -O2 -pipe -march=x86-64 -mtune=generic -O2
-pipe -DCYTHON_CLINE_IN_TRACEBACK=0 -Isrc -Isrc\lxml\includes
-ID:\nightlight\.venv\include
-ID:\Setup\MingW\installation\mingw64\include\python3.9 -c src\lxml\etree.c
-o build\temp.mingw_x86_64-3.9\src\lxml\etree.o -w
    In file included from src\lxml\etree.c:711:
    src\lxml\includes/etree_defs.h:14:10: fatal error: libxml/xmlversion.h:
No such file or directory
       14 | #include "libxml/xmlversion.h"
          |          ^~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
    Compile failed: command
'D:\\Setup\\MingW\\installation\\mingw64\\bin\\gcc.exe' failed with exit
code 1
    creating Users
    creating Users\Dell
    creating Users\Dell\AppData
    creating Users\Dell\AppData\Local
    creating Users\Dell\AppData\Local\Temp
    gcc -mdll -O2 -Wall -I/usr/include/libxml2 -c
C:\Users\Dell\AppData\Local\Temp\xmlXPathInitd23t_5hf.c -o
Users\Dell\AppData\Local\Temp\xmlXPathInitd23t_5hf.o
    C:\Users\Dell\AppData\Local\Temp\xmlXPathInitd23t_5hf.c:1:10: fatal
error: libxml/xpath.h: No such file or directory
        1 | #include "libxml/xpath.h"
          |          ^~~~~~~~~~~~~~~~
    compilation terminated.
    error: command 'D:\\Setup\\MingW\\installation\\mingw64\\bin\\gcc.exe'
failed with exit code 1

*********************************************************************************
    Could not find function xmlCheckVersion in library libxml2. Is libxml2
installed?

*********************************************************************************
    ----------------------------------------
ERROR: Command errored out with exit status 1:
'D:\nightlight\.venv\bin\python3.exe' -u -c 'import io, os, sys,
setuptools, tokenize; sys.argv[0] =
'"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-install-nbpaxmto\\lxml_f6d5c4e067ed43e8bf4a1f3a1db7bbca\\setup.py'"'"';
__file__='"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-install-nbpaxmto\\lxml_f6d5c4e067ed43e8bf4a1f3a1db7bbca\\setup.py'"'"';f
= getattr(tokenize, '"'"'open'"'"', open)(__file__) if
os.path.exists(__file__) else io.StringIO('"'"'from setuptools import
setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"',
'"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))'
install --record
'C:\Users\Dell\AppData\Local\Temp\pip-record-7xh23b80\install-record.txt'
--single-version-externally-managed --compile --install-headers
'D:\nightlight\.venv\include\site\python3.9\lxml' Check the logs for full
command output.
WARNING: You are using pip version 21.2.4; however, version 22.1.2 is
available.
You should consider upgrading via the 'D:\nightlight\.venv\bin\python3.exe
-m pip install --upgrade pip' command.
